### PR TITLE
Make plugin version optional in version catalogs

### PR DESCRIPTION
**Pull request by unshare**
In some cases, one can apply a plugin in `plugin { ... }` block without specifying a version.
However, at the moment it's not possible to leave version unspecified in a version catalog and then apply via `alias`. Under the hood `alias` calls `version` unconditionally and then fails on null or empty value. This commit prevents such a failure.

Use case for not specifying plugin versions in version catalog:
* a plugin may be applied from an included build (incl. `buildSrc`)
* plugin version may be pinned via settings, a settings plugin or an init script via `pluginManagement`
* a plugin may be discovered in build script classpath
* other situations and mechanisms unknown to me

Even if version is not specified, version catalog can be used as a bill of materials or a reference list. Besides, generated DSL accessors work as a typo prevention mechanism.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by ov7a** [comment]
Thank you for your contribution!

Please add tests to cover your changes.

**Comment by unshare** [comment]
@bot-gradle test this

**Comment by unshare** [comment]
Hello, @ov7a!
Sorry for triggering the workflows. It takes hours for me to run tests locally whereas your CI completes within minutes due to build caching.
I've taken another look at the offending code fragment, and I've realised there is undesirable code duplication: the interface has default method implementation which is okay for me, and two identical overrides that fail me and should be removed IMO.
For the time being, please review the PR as as.
Providing tests is still a work in progress on me, but frankly, it's not trivial given the current coverage.

**Comment by ov7a** [comment]
@unshare [See CONTRIBUTING.md](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md#remote-build-cache): you can benefit from the remote cache too. Also, you can select tests specific to the area of your changes, no need to re-run all tests on each iteration.

Your changes are only about removing some lines. Is it expected?

**Comment by unshare** [comment]
@ov7a Thanks for the pointer to the build cache. Too late for this engagement: I've generated my own local one. But thanks.
With regard to the delete-only nature of this PR: yes, this is my intention. The overriden implementation don't handle missing version whereas the default implementation in the interface does it correctly, IMO.
See: https://github.com/gradle/gradle/blob/30fa5bccd047c28bf63c318915bc575c7e7e32b7/subprojects/core-api/src/main/java/org/gradle/plugin/use/PluginDependenciesSpec.java#L139

**Comment by ov7a** [comment]
Okay, I asked just to clarify this and be 100% sure.
While I understand where are you coming from with that change, and it seems to be fine, please make sure that all new scenarios are also covered. See [integTests](https://github.com/gradle/gradle/tree/30fa5bccd047c28bf63c318915bc575c7e7e32b7/platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/catalog).

**Comment by cobexer** [comment]
Thank you for your contribution!

This PR has a valid DCO. The relevant team for this area will confirm the design of the implementation choices.

**Comment by big-guy** [comment]
Sorry, this isn't something we're currently working on and won't have time to take a look at it for awhile.

**Comment by patton73** [comment]
Guys, really merge this PR. Stop adding new features. Please complete already existing ones.

**Comment by sgammon** [comment]
Forgive me, I am on my phone, but:

plugins {
  id(libs.plugins.abc.get().pluginId)
}

Works fine on Gradle latest. Of course you can also expose a convention plugin which allows a direct implementation dependency.

**Comment by unshare** [comment]
@sgammon
Sorry for passing evaluative judgement, but you're defending an obviously buggy piece of code and advocating for an ugly workaround.
Please take a look at the patch itself and the conversation above.

**Comment by unshare** [comment]
Everyone, I'm sorry for the delay implementing tests. It's not a top priority for me.
Gradle have requested the test cases for ALL use cases. And the target milestone isn't in the immediate future.

One more piece of evaluative judgement, sorry: Gradle as organization behave like an average enterprise. I.e. they deny having bugs to the utmost and hesitate to accept outside contributions fearing regression.
Most of us witness such a behavior on our day jobs. We should just keep calm and carry on the constructive dialogue.

**Comment by sgammon** [comment]
@unshare I shared a workaround in case people are hitting this issue and waiting on your PR. Good luck getting it merged.

**Comment by sgammon** [comment]
Maybe focus your wrath on Gradle instead of other users who are browsing in your PR and trying to be helpful.

I don't think they are going to merge it.

**Comment by mavek87** [comment]
@sgammon No it does NOT work. I have this in a precompiled script plugin ```plugins {
//    id "org.hidetake.ssh"
    id(libs.plugins.gradle.ssh.plugin.get().pluginId)
}``` and even though I'm using a catalog your workaround doesn't work with groovy ```
FAILURE: Build failed with an exception.

* Where:
Script '/home/mavek/Documenti/workspace/java/rebirth-gradle-deploy-plugin/src/main/groovy/it.rebirthproject.rebirth-ssh-deploy.gradle' line: 3

* What went wrong:
Could not compile script '/home/mavek/Documenti/workspace/java/rebirth-gradle-deploy-plugin/src/main/groovy/it.rebirthproject.rebirth-ssh-deploy.gradle'.
> startup failed:
  script '/home/mavek/Documenti/workspace/java/rebirth-gradle-deploy-plugin/src/main/groovy/it.rebirthproject.rebirth-ssh-deploy.gradle': 3: argument list must be exactly 1 literal String or String with property replacement
  
  For more information on the plugins {} block, please refer to https://docs.gradle.org/8.6/userguide/plugins.html#sec:plugins_block in the Gradle documentation.
  
   @ line 3, column 5.
         id(libs.plugins.gradle.ssh.plugin.get().pluginId)
         ^
  
  1 error
```

**Comment by sgammon** [comment]
@mavek87 Where in particular is it failing to find a symbol? I can't tell from your comment; if you fix the code blocks I may be able to help.

**Comment by mavek87** [comment]
@sgammon now the code in the comment should be fixed

**Comment by sgammon** [comment]
@mavek87 I can't quite tell what's going on, but here are the steps I would try:

**1) If you are in a `buildSrc` project**

You will need to make your catalog available to your precompiled plugins. You can do this with:

```kotlin
dependencies {
  implementation(files(libs.javaClass.superclass.protectionDomain.codeSource.location))
}
```
> This assumes your version catalog symbol is `libs`. If it's, for example, `other`, you would do `files(other.javaClass...`.

This fix is from the Gradle team; you can read more [here](https://github.com/gradle/gradle/issues/15383)

**2) If you are in a precompiled Kotlin plugin, you may need to create the accessor**

You can try this with:

```kotlin
import org.gradle.accessors.dm.LibrariesForLibs
// ...
val libs = the<LibrariesForLibs>()

// now you can access `libs.x.y.z` etc
```

**3) If the error is in your IDE, you may need to suppress DSL scope violations**

You can do this with:
```kotlin
@file:Suppress("DSL_SCOPE_VIOLATION")
```

These code samples are in Kotlin, I'm sorry, I'm not very good at Groovy. But they should work roughly the same way. It seems you are building a Gradle plug-in, right? If you can access `libs`, you should be able to do this trick because

```
libs.plugins.gradle.ssh.plugin.get().pluginId
```

Should be returning a string. Here's what I get for a similar plugin in my IDE:
Screenshot 2024-03-06 at 11 14 34 AM

For the following mapping in the catalog:
```toml
[versions]
kotlin-sdk = "1.9.22"
# ...

[plugins]
kotlin-multiplatform = { id = "org.jetbrains.kotlin.multiplatform", version.ref = "kotlin-sdk" }
# ...
```

I am running Gradle 8.6 and 8.7-RC2 in this case.

**Comment by sgammon** [comment]
Oh, one other thing. If you have catalog structures that are nested, you may need to use a provider. For example

```toml
[versions]
# ...

[plugins]
example-plugin = { ... }
example-plugin-nested = { ... }
```

To access `example-plugin`'s plugin ID with `libs`, sometimes I need to do:

```kotlin
libs.plugins.example.plugin.asProvider().get().pluginId
```

**Comment by JavierSegoviaCordoba** [comment]
> @sgammon No it does NOT work. I have this in a precompiled script plugin ```plugins {
> 
> //    id "org.hidetake.ssh"
> 
>     id(libs.plugins.gradle.ssh.plugin.get().pluginId)
> 
> }``` and even though I'm using a catalog your workaround doesn't work with groovy 
> 
> 
> 
> ```
> 
> FAILURE: Build failed with an exception.
> 
> 
> 
> * Where:
> 
> Script '~/Documenti/workspace/java/rebirth-gradle-deploy-plugin/src/main/groovy/it.rebirthproject.rebirth-ssh-deploy.gradle' line: 3
> 
> 
> 
> * What went wrong:
> 
> Could not compile script '~/Documenti/workspace/java/rebirth-gradle-deploy-plugin/src/main/groovy/it.rebirthproject.rebirth-ssh-deploy.gradle'.
> 
> > startup failed:
> 
>   script '~/Documenti/workspace/java/rebirth-gradle-deploy-plugin/src/main/groovy/it.rebirthproject.rebirth-ssh-deploy.gradle': 3: argument list must be exactly 1 literal String or String with property replacement
> 
>   
> 
>   For more information on the plugins {} block, please refer to https://docs.gradle.org/8.6/userguide/plugins.html#sec:plugins_block in the Gradle documentation.
> 
>   
> 
>    @ line 3, column 5.
> 
>          id(libs.plugins.gradle.ssh.plugin.get().pluginId)
> 
>          ^
> 
>   
> 
>   1 error
> 
> ```

Your problem is totally unrelated to the PR and the workaround provided, which indeed, works.

Look for version catalogs in included builds and how the plugin block works.

**Comment by sgammon** [comment]
@JavierSegoviaCordoba The workaround is related because it lets you select an existing plugin ID or version, as a string, without repeating yourself. You can pull it from the catalog, even in an included build. It obviously depends where the consuming build _is_ on disk, because you may need to add the catalog to `settings.gradle[.kts]`, etc.

The workaround works for me. I fear Gradle will not merge this because it means specifying no plugin version in the catalog, which may be re-used by other projects. In the interest of actually unblocking developers, I'm just trying to help, but thank you for the pointers.

**Comment by JavierSegoviaCordoba** [comment]
> @JavierSegoviaCordoba The workaround is related because it lets you select an existing plugin ID or version, as a string, without repeating yourself. You can pull it from the catalog, even in an included build. It obviously depends where the consuming build _is_ on disk, because you may need to add the catalog to `settings.gradle[.kts]`, etc.
> 
> 
> 
> The workaround works for me. I fear Gradle will not merge this because it means specifying no plugin version in the catalog, which may be re-used by other projects. In the interest of actually unblocking developers, I'm just trying to help, but thank you for the pointers.

It is totally unrelated.

The `plugins` block is extracted, and analyzed separately, so it is impossible to use a catalog inside the `plugins` block in a precompiled script plugin in an included build or in the sources of a regular project.

It will not work, neither `id` or `alias`, with or without this PR being merged.

**Comment by sgammon** [comment]
> The plugins block is extracted, and analyzed separately, so it is impossible to use a catalog inside the plugins block in a precompiled script plugin in an included build or in the sources of a regular project.

It is not **without [this workaround](https://github.com/gradle/gradle/issues/15383)** from the Gradle team. I am accessing my catalogs fine in my precompiled scripts.

**Comment by sgammon** [comment]
@JavierSegoviaCordoba Technically you can circumvent use of the `plugins { }` block in any capacity, at all, by simply providing a plugin implementation of some kind on the classpath. Then such plugins can be invoked with `id(...)` or even `apply(plugin = ...)`, as you probably know. My workaround is related because it allows re-use of the ID from an existing catalog. This can be accomplished in different ways from all the places you named above, just not gracefully.

I don't actually know why you are attacking my workaround. At the worst case it won't work, sorry. At the best case someone can find some use for it. Gradle can be very confusing even for experts; there is no sense in us fighting about an issue that we are not empowered to fix.

**Comment by JavierSegoviaCordoba** [comment]
Again, it will not work. Read all posts in the issue you shared, the author of the hack-workaround explicitly says it does not work inside `plugins` block.

And `apply` or any variant like `pluginManager` does not work in the same way `plugins` block works, they are not extracted ahead in order to generate any accessor like `plugins` block does. If you uses the `plugins` block in a precompiled script plugin and you apply the kotlin plugin, you will see all the accessors for the Kotlin extension, `kotlin`, the compile tasks, `tasks.compileKotlin` and so on. That only happens due how `plugins` block works.

So this PR will not fix that issue, but it is not even aiming to fix that issue as it is unrelated. And the hack-workaround is limited and will not be able to work inside `plugins` until Gradle fixes the other issue. Not sure if it will be even possible to fix the part of the `plugins` block due a chicken and egg situation.

**Comment by sgammon** [comment]
> Again, it will not work.

To do what, fix the PR? I never claimed it would. People arrive here searching Google for different errors. I did, right?

> the author of the hack-workaround explicitly says it does not work inside plugins block.

Thanks, I addressed that

> And apply or any variant like pluginManager does not work in the same way plugins block works, they are not extracted ahead in order to generate any accessor like plugins block does

Yes, it sounds like we will have to live without accessors until Gradle merges this and perhaps some other fixes. Accessors are cosmetic

> you will see all the accessors for the Kotlin extension

```kotlin
the<KotlinProjectExtension>().apply {
  // ...
}
```

> ... the compile tasks, `tasks.compileKotlin`...

```kotlin
tasks.named("compileKotlin", KotlinCompile::class).configure {
  // ...
}
```

> ...not even aiming to fix that issue as it is unrelated...

From OP (bolding mine):
In some cases, one can apply a plugin in `plugin { ... }` block without specifying a version.
However, at the moment it's not possible to leave version unspecified in a version catalog and **then apply via alias**

The workaround I have used for this case in the past is the one I posted. Relevance is subjective, anyway. Do you work for Gradle? Maybe JetBrains? If not, why are you here? To tell me my workaround, offered in good faith, was bad? Okay.

> Not sure if it will be even possible to fix the part of the plugins block due a chicken and egg situation.

I still don't see an issue. A plugin is an implementation on the classpath, addressable by a plugin ID, which resides at a known spot on disk. You then apply a plugin via various means, the `plugins { ... }` block is just one of these.

> ... PD: I am not attacking your workaround, indeed I am defending it...

Also you:

> Your problem is totally unrelated to the PR
> It is totally unrelated

> ... my reply was to the person who said your workaround...

You might want to tag who you are talking to because my comment is immediately above yours.

**Comment by mavek87** [comment]
@sgammon This is my project. Gradle init and choose always groovy. No buildSrc, nothing. Just one included project (named "plugin")

this is my settings gradle in the root

```
pluginManagement {
    repositories {
        gradlePluginPortal()
    }
}

dependencyResolutionManagement {
    versionCatalogs {
        libs {
            plugin('gradle-ssh-plugin', 'org.hidetake.ssh').version("2.11.2")
        }
    }
}

rootProject.name = 'example-gradle'
include('plugin')
```

image

In the build.gradle of the "plugin" subproject i have just 

```
plugins {
       id 'groovy-gradle-plugin'
}
```

And in the precompiled script I have just this

```
plugins {
//    id "org.hidetake.ssh"
    id(libs.plugins.gradle.ssh.plugin.get().pluginId)
}
```

image

If I use the string it works. If I try with the "workaround" I receive the same error I posted you before.

**Comment by sgammon** [comment]
@mavek87 @JavierSegoviaCordoba is right that your issue may be different; if you post it in a new issue (I'm sorry to ask, but it will be cleaner and friendlier to ppl here), I'm happy to give a shot helping out. I'll review the code you've posted here as well.

**Comment by JavierSegoviaCordoba** [comment]
@sgammon I didn't tag you, I quoted the other person message who said the workaround you shared does not work, and I said it works. For some reason, you replied me.

The rest of tags is to explain why the workaround will work in any situation except the `plugins` block in precompiled plugins and why it is unrelated to this PR.

I still do not understand why being able to apply a plugin without a version is related to a hack-workaround to use catalogs in any place except `plugins` block in precompiled script plugins (with or without version, none works), but for some reason, you pursue that goal.

**Comment by sgammon** [comment]
@JavierSegoviaCordoba When you are replying to an email from GitHub, it will enclose all sorts of context; so I have no way of knowing if you are meaning to quote him or not.

Yes, I see now that I was confused. I appreciate that you are not attacking the workaround I posted.

I encouraged this Gradle user to file another bug so we can discuss it there without polluting this discussion.

Ultimately I just want everyone's Gradle builds to work smoothly. Sorry for misidentifying your intentions and tone. In the future, it would help someone in my position to clearly tag who you are talking to.

**Comment by sgammon** [comment]
> I still do not understand why being able to apply a plugin without a version is related to a hack-workaround to use catalogs in any place except plugins block

I do not presume to know every reason why a developer wants to do something. If they want a catalog somewhere, there are usually ways to do it. Catalogs are useful and help with less repetition.

I don't see any problem and I don't think your understanding is needed.

**Comment by JavierSegoviaCordoba** [comment]
@sgammon that clarifies the misunderstood. To be honest I haven't replied via email in GitHub before so I didn't have any clue about that :)

**Comment by sgammon** [comment]
@JavierSegoviaCordoba No problem, sorry as well. Neither of us work for Gradle or own this PR, though, so we should probably continue discussing elsewhere if needed. Cheers.

**Comment by mavek87** [comment]
@JavierSegoviaCordoba the reason is because I dont want to type and duplicate something I also have in my catalog yet ("org.hidetake.ssh")

**Comment by sgammon** [comment]
@mavek87 you could name it something like "Issues with Version Catalog accessors for precompiled plugin" and tag me.

**Comment by JavierSegoviaCordoba** [comment]
> @JavierSegoviaCordoba the reason is because I dont want to type and duplicate something I also have in my catalog yet ("org.hidetake.ssh")

The only way is not using `plugins` block. I am not a groovy user (my knowledge about it is almost zero), maybe using `apply` or applying the plugins with `pluginManager` does not force you to access the extensions as you would do in Kotlin with `the` or `configure`. If not, you will need to use those methods as you would be losing the accessors in the precompiled plugin.

**Comment by sgammon** [comment]
@JavierSegoviaCordoba Groovy has dynamic property resolution features that Kotlin doesn't have, so it is entirely possible that the accessors do work in Groovy. Groovy can, in some cases, also reach across `private` boundaries. Spock does this for testing. So I would not actually bet on that (this is coming from another heavy Kotlin user, I don't know for sure).

**Comment by mavek87** [comment]
@sgammon I opened a new issue: https://github.com/gradle/gradle/issues/28371

Unfortunately I'm the opposite, I really dont understand nothing about Kotlin. Trying for days to look in the documentation for this and other problems I'm facing but it's really though

**Review by tresat — APPROVED** [review]

**Review by octylFractal — APPROVED** [review]

**Review by lkasso — APPROVED** [review]
Docs LGTM.

**Pull request merged**
